pkg/oci/oci.go

```diff
@@ -0,0 +1,316 @@
+	if err != nil {
+		return nil, fmt.Errorf("get definition from descriptor: %w", err)
+	}
+
+	return definition, nil
+}
+
+func GetEbpfProgram(target oras.Target, authOpts *AuthOptions, image string) ([]byte, error) {
+	err := PullIfNotExist(target, authOpts, image)
+	if err != nil {
+		return nil, fmt.Errorf("pull image: %w", err)
+	}
+	index, err := GetImageListDescriptor(target, image)
+	if err != nil {
+		return nil, fmt.Errorf("get image list descriptor: %w", err)
+	}
+	manifest, err := GetHostArchManifest(target, index)
+	if err != nil {
+		return nil, fmt.Errorf("get arch manifest: %w", err)
+	}
+	if len(manifest.Layers) != 1 {
+		return nil, fmt.Errorf("expected exactly one layer, got %d", len(manifest.Layers))
+	}
+	definition, err := GetContentFromDescriptor(target, manifest.Layers[0])
+	if err != nil {
+		return nil, fmt.Errorf("get ebpf program from descriptor: %w", err)
+	}
+
+	return definition, nil
+}
+
+func GetTagFromImage(image string) (string, error) {
+	repo, err := reference.Parse(image)
+	if err != nil {
+		return "", fmt.Errorf("parse image %q: %w", image, err)
+	}
+	tagged, ok := repo.(reference.Tagged)
+	if !ok {
+		return "latest", nil
+	}
+	return tagged.Tag(), nil
+}
+
+func GetRepositoryFromImage(image string) (string, error) {
+	repo, err := reference.Parse(image)
+	if err != nil {
+		return "", fmt.Errorf("parse image %q: %w", image, err)
+	}
+	if named, ok := repo.(reference.Named); ok {
+		return named.Name(), nil
+	}
+	return "", fmt.Errorf("image has to be a named reference")
+}
+
+func NormalizeImage(image string) (string, error) {
+	name, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return "", fmt.Errorf("parse normalized image %q: %w", image, err)
+	}
+	return reference.TagNameOnly(name).String(), nil
+}
+
+func GetHostString(repository string) (string, error) {
+	repo, err := reference.Parse(repository)
+	if err != nil {
+		return "", fmt.Errorf("parse repository %q: %w", repository, err)
+	}
+	if named, ok := repo.(reference.Named); ok {
+		return reference.Domain(named), nil
+	}
+	return "", fmt.Errorf("image has to be a named reference")
+}
+
+func NewAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Client, error) {
+	logrus.Debugf("Using auth file %q", authOptions.AuthFile)
+
+	var cfg *configfile.ConfigFile
+	var err error
+
+	// 1. Explicitly setting the auth file
+	// 2. Using the default auth file
+	// 3. Using the default docker auth file if 2. doesn't exist
+	if authOptions.AuthFile != DefaultAuthFile {
+		authFileReader, err := os.Open(authOptions.AuthFile)
+		if err != nil {
+			return nil, fmt.Errorf("open auth file %q: %w", authOptions.AuthFile, err)
+		}
+		defer authFileReader.Close()
+		cfg, err = config.LoadFromReader(authFileReader)
+		if err != nil {
+			return nil, fmt.Errorf("load auth config: %w", err)
+		}
+	} else if _, err := os.Stat(authOptions.AuthFile); err == nil {
+		authFileReader, err := os.Open(authOptions.AuthFile)
```

Comment:
We could use `os.Open` directly here instead of calling `os.Stat` first.

Reply:
xref https://github.com/inspektor-gadget/inspektor-gadget/pull/2058#discussion_r1332738001. 

Updated to use my suggestion there and @burak-ok's comment.